### DISCUSSION TITLE
Port tty-clock to ELKS

### DISCUSSION THREAD

**Discussion by toncho11** [jbruchon/elks#1817] [Ideas]
https://github.com/xorg62/tty-clock

I think @ghaerr you will like this? We have a ncurses replacement, right?

**Comment by ghaerr** [comment]
Hello @toncho11,

Yes, looking quickly at `tty-clock`, it definitely uses more of our super-simple ncurses library than we have implemented, but even without enhancing ncurses could probably be slightly rewritten and ported to ELKS in a fairly straightforward fashion.

Did you happen to notice the author's other works? There's a [tty-tetris](https://github.com/xorg62/tty-tetris) that I'm sure you'd like as well as a [tty-pong](https://github.com/xorg62/tty-pong) game, and tty-tetris doesn't even require ncurses.

I compiled up both on macOS without modifications. Here's tty-tetris:
<img width="801" alt="tty-tetris" src="https://github.com/ghaerr/elks/assets/11985637/1471c1d3-7182-411f-9787-00c248de41a1">

**Reply by toncho11** [reply]
Nice!
The three of them will be fun additions and demonstrators for ELKS. I can help when I have a bit more time.

**Comment by tyama501** [comment]
Hello @ghaerr ,

Last year when I worked for porting moria I added some functions to the curses, but I forgot to share with you.
https://github.com/tyama501/ELKSmoria/blob/elksmoria/elks/curses.c
https://github.com/tyama501/ELKSmoria/blob/elksmoria/elks/curses.h

Can I merge this? 
Or would you like to keep it as minimum as possible for this repository?
(Most of them are empty, and only mvaddstr and clear are functional. So only these two may be enough.
 And I noticed *newwin is not good returning local pointer, I need to fix it)

Other than curses,
I remember there are still many bugs in ELKSmoria I need to fix...

**Reply by ghaerr** [reply]
Hello @tyama501,

I would say to hold off at this point. The main reasons are 1) returning even a valid pointer from `newwin` means ncurses will have to keep a background screen/window representation for everything passed to ncurses, and fully supporting ncurses windows probably isn't worth it, and 2) `wrefresh` and friends won't work. My initial port (and it looks like yours too) was just to barely scrape by with cursor position handling on the main screen without having to rewrite a ported program much.

If you'd like to take a pass at `tty-clock` or `tty-tetris`, I would suggest replacing the ncurses calls creating windows and outputting to them with just outputting to to the main screen using the existing tiny ncurses (just like `sl` does). In this way, a very simple ncurses will still work. I notice in `tty-clock` that the only reason a window is used is to draw a PC-graphics-line style box around the clock - something that could be done directly in tty-clock.c instead of curses, since there is no requirement for overlapping windows.

Thank you!

**Reply by tyama501** [reply]
OK, thanks.

**Comment by tyama501** [comment]
Reference
https://github.com/ghaerr/elks/pull/1855

https://github.com/ghaerr/elks/pull/1858

https://github.com/ghaerr/elks/pull/1864

**Comment by tyama501** [comment]
ttyclock worked on pc-98 emulater with a one-line modification.

https://github.com/tyama501/elks/commit/51d3cbd6ac080fd92f86734538f5100bad7f438c

![ttyclock_pc98](https://github.com/ghaerr/elks/assets/61556504/4724051d-f759-4350-a694-bf2264d2b031)

**Comment by ghaerr** [comment]
Fantastic! Definitely send over a PR. You can test cycle through the colors by typing 1-7, (0=black), and also select seconds or not with 's'.

**Reply by tyama501** [reply]
![image](https://github.com/ghaerr/elks/assets/61556504/7002546f-5ae6-405a-869c-962c57bb447e)

Color looks good except 7.
7 should be white right?

![image](https://github.com/ghaerr/elks/assets/61556504/97b76c24-588a-4d69-a3e1-b4e98b03bd56)

And the second is shown even -s is not set...

Well, I think good enough for now so I will make PR.

**Comment by tyama501** [comment]
Pong (demo?)
![image](https://github.com/ghaerr/elks/assets/61556504/5f0f867d-954e-4e6d-9b5e-125ae8044b36)

**Comment by tyama501** [comment]
ttytetris

Maybe this needs white. (Frame and white blocks are not shown)
And it is very slow.
I recommend nxtetris.

![image](https://github.com/ghaerr/elks/assets/61556504/ea56c334-9158-44ca-871a-c75e92cfb7f2)

**Reply by ghaerr** [reply]
> Color looks good except 7.
7 should be white right?

Yes. It looks like color 7 is not working properly on PC-98 console.

> ttytetris
Maybe this needs white. (Frame and white blocks are not shown)

Yes, I think same problem.

> And it is very slow.

More investigation needed. Sounds like new system call `setitimer` is not working properly on PC-98. This allows for sub-second timeouts, specified in microseconds. You can adjust and play with this by changing TIMING variable in ttytetris.h.

> Pong (demo?)

Yes, its not really a game. How is the speed, is it OK?

> I think good enough for now so I will make PR.

Go ahead, the white values can be fixed later, I suppose. BTW, this could be due to 7 being 0 in reverse video mode, and then compare with zero you recently added/changed.

I have a PR coming which optimizes tui/curses.c coming for smaller executables but that won't affect this work.

**Reply by ghaerr** [reply]
All the tty-curses apps including ttyclock and ttytetris reverse video (ESC [ 7 m) followed by SPACE to get a white background, in contrast to colored backgrounds where the set fg/bg color ANSI sequence ESC [ fg; bg m is used, where fg is 30 through 37 and bg 40 through 47. So the PC-98 color conversion routine(s) may need to look both bg and fg color(s) in order to work.

**Reply by ghaerr** [reply]
> And it is very slow.
> More investigation needed. Sounds like new system call setitimer is not working properly on PC-98.

On second thought, I believe the problem is related to the already-explained ANSI cursor and color sequences being slow to interpret by ELKS, and the setitimer function being used to delay after an already slow process. What is needed in all these style programs is a real-time delay (or none) that calculates a delay based on the real time elapsed between drawing start/finish and subtracting that from the "normal" delay such that the display cycle time is constant. 

For the time being, only the ttypong program can be passed an argument specifying a fixed delay. Ttytetris would have to be recompiled changing the `TIMING` variable in ttytetris.h. 

The `sl` program does calculate a delay based on real time, I suppose this needs to be added to the tty games.

**Reply by tyama501** [reply]
Hello @ghaerr, 

It seems that fg == bg when color is 7.
I modified no display condition.
https://github.com/tyama501/elks/commit/5dc3260c0ee4a620ce1b7580f0ae038682411769

Now the white worked.
![ttyclock_white](https://github.com/ghaerr/elks/assets/61556504/3f2eacb5-4d75-4052-8c92-fee7f7d4ea05)

Maybe I thought the ttytetris is slow because the white block was not shown.
I dont't feel that much slow any more on the 486SX, 25MHz emulator.

https://github.com/ghaerr/elks/assets/61556504/57f51722-b45f-4f3d-8070-5d946db286ac

I will make another PR.

**Comment by ghaerr** [comment]
> It seems that fg == bg when color is 7.
I modified no display condition.
> Now the white worked.

Nice!!

> Maybe I thought the ttytetris is slow because the white block was not shown.
I dont't feel that much slow any more on the 486SX, 25MHz emulator.

Thank you for the input. I would agree perceived slowness would seem higher when no white blocks shown. I'll hold off on calculated cycle delay for the time being. Ttytetris should be delaying 300ms between each block drop. The game speeds up the longer it's played.
